### DISCUSSION TITLE
LinkTokenMock -- getting OverFlow/UnderFlow Error

### DISCUSSION THREAD

**Discussion by DeanKrupka** [Cyfrin/foundry-full-course-cu#1179] [Q&A] [answered]
I am currently doing a bit of extra-curricular study -- trying to combine some lessons. The functionality I'm going for is a contract able to accept any ERC20 token as donation. I've written a test to see if that functionality works (called 'testTossToken' at bottom, however I'm repeatedly facing an error: 'Arithmetic over/underflow.' I would very much appreciate any assistance. 

My function in my main contract looks like this:
~~~
    function approveTossToken(address _tokenAddress, uint256 _value) public {
            bool approveToss = ERC20(_tokenAddress).approve(address(this), _value);
            require(approveToss, "ERC20 Approval failed.");
    }

    function tossToken(address _tokenAddress, uint256 _value) public {
        //Toss - Perform Transfer
        if (_value == 0) revert Fountain__ValueMustBeGreaterThanZero();
        bool transferSuccess = ERC20(_tokenAddress).transferFrom(
            msg.sender,
            address(this),
            _value
        );
        require(transferSuccess, "ERC20 Toss failed.");

        //Accounting of Tosses
        s_tossers.push(msg.sender);
        s_tokenAddresses.push(_tokenAddress);
        tokenAddressToTosses[_tokenAddress].push(
            TosserAddressAndAmount(msg.sender, _value)
        );
        tokenAddressToTotalTossed[_tokenAddress] += _value;

        emit TossedEthOrToken(msg.sender, _value, _tokenAddress);
    }
~~~

My test looks like this. I included the setUp() function as well as my state variables in case they are relevant:

~~~
    uint256 public constant ETHTOSS_AMOUNT = 0.5 ether;
    uint256 public constant TOKENTOSS_AMOUNT = 0.5 ether; // Actually Link

    Fountain fountain;
    LinkTokenMock linkTokenMock;
    uint256 public constant STARTING_USER_BALANCE = 10 ether;
    uint256 public constant STARTING_USERTOKEN_BALANCE = 9 ether;
    address USER = makeAddr("user");
    address USER2 = makeAddr("user2");
    address USER3 = makeAddr("user3");

    function setUp() public {
        DeployFountain deployFountain = new DeployFountain();
        fountain = deployFountain.run();

        linkTokenMock = new LinkTokenMock();
        linkTokenMock.transfer(USER, STARTING_USERTOKEN_BALANCE);
        linkTokenMock.transfer(USER2, STARTING_USERTOKEN_BALANCE);
        linkTokenMock.transfer(USER3, STARTING_USERTOKEN_BALANCE);

        vm.deal(USER, STARTING_USER_BALANCE);
        vm.deal(USER2, STARTING_USER_BALANCE);
        vm.deal(USER3, STARTING_USER_BALANCE);
    }

    function testCheckSetUp() public {
        assertEq(STARTING_USERTOKEN_BALANCE, linkTokenMock.balanceOf(USER));          //THIS TEST PASSES
    }
    
    function testTossToken() public {
        //Arrange
        //Act
        vm.prank(USER);
        fountain.tossToken(address(linkTokenMock), TOKENTOSS_AMOUNT);
        //Assert
        // assertEq(fountain.getTossers(1), USER);
        // assertEq(fountain.getTotalTossedByTokenAddress(0), TOKENTOSS_AMOUNT);
      }
~~~
        
The final function call:

~~~
        fountain.tossToken(address(linkTokenMock), TOKENTOSS_AMOUNT);
~~~

 is where I am getting the error. If anyone can offer advice, I would greatly appreciate it. Thank you!

**Comment by DeanKrupka** [comment]
Ok, so I have figured out that ~somehow~ it is the approval of the ERC20 step that is the issue. But still cant figure out why. Below is the test for Approval:
```
    function testApproveTossToken() public {
        //Arrange
        //Act
        vm.prank(USER);
        fountain.approveTossToken(address(linkTokenMock), TOKENTOSS_AMOUNT);
        console.log("USER's Link Balance:", linkTokenMock.balanceOf(USER));
        console.log(
            "Fountain's Link Balance:",
            linkTokenMock.balanceOf(address(fountain))
        );
        console.log(
            "Fountain's Link Allowance for USER:",
            linkTokenMock.allowance(USER, address(fountain))
        );
    }
```
And the logged results HERE:
```
[PASS] testApproveTossToken() (gas: 54757)
Logs:
  USER's Link Balance: 9000000000000000000
  Fountain's Link Balance: 0
  Fountain's Link Allowance for USER: 0
```

**Reply by kingofclubstroyDev** [reply, marked as answer]
So the way ERC20 token approvals work is that the caller of the approve() function approves the passed in address for the amount specified. In the case of your approveTossToken() function, the caller is the contract, and the address to approve is also the contract, so it is approving itself which is not what you want. So what needs to be done is the user needs to call approve() directly on the token contract and pass is the contracts address.

In your test it should be something like:

 function testApproveTossToken() public {
        //Arrange
        //Act
        vm.prank(USER);
        // notice the call is to the link mock contract, and the address is the fountain contract
        linkTokenMock.approve(address(fountain), TOKENTOSS_AMOUNT);
        console.log("USER's Link Balance:", linkTokenMock.balanceOf(USER));
        console.log(
            "Fountain's Link Balance:",
            linkTokenMock.balanceOf(address(fountain))
        );
        console.log(
            "Fountain's Link Allowance for USER:",
            linkTokenMock.allowance(USER, address(fountain))
        );
    }

**Reply by DeanKrupka** [reply]
Amazing. I knew it was something fundamental I was missing. My mental model for how msg.sender works, and more generally where this function call needed to originate from needed updating.  Thanks very much.
